app/services/badges/award_thumbs_up.rb

```diff
@@ -0,0 +1,48 @@
+module Badges
+  class AwardThumbsUp
+    THUMBS_UP_BADGES = {
+      1 => "100 Thumbs Up Milestone",
+      2 => "500 Thumbs Up Milestone",
+      3 => "1,000 Thumbs Up Milestone",
+      4 => "5,000 Thumbs Up Milestone",
+      5 => "10,000 Thumbs Up Milestone"
+    }.freeze
+
+    MIN_THRESHOLD = THUMBS_UP_BADGES.keys.min
+
+    def self.call
+      badge_ids = fetch_badge_ids
+      user_thumbsup_counts = Reaction
+        .where(category: "thumbsup", reactable_type: "Article")
+        .group(:user_id)
+        .having("COUNT(*) >= ?", MIN_THRESHOLD)
+        .order(Arel.sql("COUNT(*) DESC"))
+        .count
+
+      user_thumbsup_counts.each do |user_id, count|
+        THUMBS_UP_BADGES.each do |threshold, _|
+          break unless count >= threshold
+
+          badge_id = badge_ids[threshold]
+          next unless badge_id
+          next unless (user = User.find_by(id: user_id))
+
+          user.badge_achievements.create(
```

Comment:
At some point before awarding we should check if it was awarded already.

Reply:
The model invalidates the awarding of these, due to the uniqueness constraint. This saves a query and is the model used by the other badge awarding code I took this approach from.

Reply:
Yup, that logic is correct and we rely on it elsewhere. In general I'd say "saving a query" isn't necessarily important for jobs which run in batches like this, but I think everything is fine in this logic.